wip: cors documentation (#1044)
apollographql/router: docs/source/configuration/cors.mdx

## Choosing CORS options for your project

When thinking about configuring CORS for your application, there are two main settings to consider:

* Which origins can access your server's resources
* Whether your server accepts user credentials (i.e., cookies) with requests

### Specifying origins

CORS uses [specific HTTP request headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#the_http_response_headers) as part of its protocol, including [`Access-Control-Allow-Origin`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin). The `Access-Control-Allow-Origin` header (ACAO) enables a server to dictate which origins can use scripts to access that server's resources.

Depending on what you're building, the origins you specify in your CORS configuration might need to change when you're ready to deploy your application.

#### Applications on private networks

If your browser is running your API on a private network (i.e., not on the public internet) and it relies on the privacy of that network for security, we strongly recommend [specifying which origins](#configuring-cors-options-for-apollo-server) can access your server's resources.

> If you instead rely on `allow_any_origin`, while your personal computer is on your private network, a script on any website could potentially make your browser talk to your private API. [Some browsers, such as Chrome, have features under development](https://wicg.github.io/private-network-access/) to solve this problem. But in the meantime, all servers on private networks should *always* specify origins in their CORS configuration.

Reviewer: Indent with `>` as this paragraph feels like a warning.

Author: fixed in 71bf26c8

#### Public or embedded APIs

If you create a public API or an API to embed in websites you don't control yourself, you probably want to allow _any_ origin to access your server's resources. You can set `allow_any_origin` in your `router.yml` configuration:

```yml
server:
  cors:
    allow_any_origin: true
```

> Using `allow_any_origin` makes the router send the [wildcard (`*`)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin#directives) value for the ACAO header, which enables _any_ website to tell a user's browser to send an arbitrary request (without cookies or other credentials) to your server and read that server's response.

#### APIs with non browser consumers

The default Apollo Server CORS behavior should suit your use case, and https://studio.apollographql.com/sandbox/explorer allows you to run queries against your Apollo Router.
You can always specify origins in your CORS configuration later. 

##  Configuring CORS options for Apollo Server

> By default, the apollo router only allows requests from https://studio.apollographql.com. If your Apollo Router is serving data for web applications, you want to either add their origin to the list of allowed origins, or switch to `allow_any_origin`.

Reviewer: ```suggestion
By default, the Apollo Router only allows requests from https://studio.apollographql.com. If your Apollo Router is serving data for web applications, you want to either add their origin to the list of allowed origins, or switch to `allow_any_origin`.
```

Author: fixed in 71bf26c8

> If you need to pass credentials to your server (e.g., via cookies), you can't use the `allow_any_origin` flag. You _must_ provide specific origins. For examples, see [Passing credentials with CORS](#passing-credentials-with-cors).

> The batteries-included `apollo-server` [has a `cors` option](../api/apollo-server/#cors), which you can use to change your server's default CORS behavior:

Reviewer: `[has a `cors` option](../api/apollo-server/#cors)` is a 404 in the preview. Maybe use the absolute URL.

Author: Good catch! this should have been removed!

Author: fixed in 71bf26c8

```yml
server:
  cors:
    origins: # ...or set a list of allowed origins
      - https://www.your-app.example.com/
      - https://studio.apollographql.com/ # So the Apollo studio can still run queries against your router
```

The example above enables CORS requests from `https://www.your-app.example`, along with `https://studio.apollographql.com`.

> Note if you plan to use [Apollo Studio Explorer](https://www.apollographql.com/docs/studio/explorer/explorer/) as a GraphQL web IDE you should include `https://studio.apollographql.com` in your list of valid origins.

> If you want to disable CORS entirely, set origins to an empty list:

Reviewer: ```suggestion
To disable CORS entirely, set origins to an empty list:
```

Author: fixed in 71bf26c8

```yml
server:
  cors:
    origins: []
```

### Passing credentials with CORS

If your server requires requests to [include a user's credentials](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#requests_with_credentials) (e.g., cookies), you need to modify your CORS configuration to tell the browser those credentials are allowed.

You can enable credentials with CORS by setting the [`Access-Control-Allow-Credentials`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials) HTTP header to `true`.

> You *must* [specify an origin](#configuring-cors-options-for-apollo-server) to enable credentialed requests. If your server sends the `*` wildcard value for the `Access-Control-Allow-Origin` HTTP header, your browser will refuse to send credentials.

> To  enable your browser to pass credentials with the batteries-included `apollo-server` you can use the [`cors` option](../api/apollo-server/#cors) to specify `origins` and set `credentials` to `true`:

Reviewer: ```suggestion
To enable your browser to pass credentials with the batteries-included `apollo-server` use the [`cors` option](../api/apollo-server/#cors) to specify `origins` and set `allow_credentials` to `true`:
```

Author: fixed in 71bf26c8
